### PR TITLE
Non-empty dirs should not be listed as objects

### PR DESCRIPTION
**Pull request by krishnasrinivas**
## Description

One interesting issue found also reported by a community member
here is the bucket has only one content A/B

```
aws --endpoint-url http://localhost:9000 --profile minio s3api list-objects --bucket testbucket3 --marker A/C 
{
    "Name": "testbucket3", 
    "Delimiter": "", 
    "MaxKeys": 10000, 
    "Prefix": "", 
    "Marker": "A/C", 
    "EncodingType": "url", 
    "IsTruncated": false, 
    "Contents": [
        {
            "LastModified": "2020-03-11T20:15:20.514Z", 
            "ETag": "\"d41d8cd98f00b204e9800998ecf8427e\"", 
            "StorageClass": "STANDARD", 
            "Key": "A/", 
            "Owner": {
                "DisplayName": "", 
                "ID": "02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4"
            }, 
            "Size": 0
        }
    ]
}
```

vs aws:
```
aws s3api list-objects --bucket harshavardhana1 --marker A/C
{
    "MaxKeys": 1000, 
    "Prefix": "", 
    "Name": "harshavardhana1", 
    "Marker": "A/C", 
    "EncodingType": "url", 
    "IsTruncated": false
}
```

## How to test this PR?

Have a single file "A/B" in a bucket and test using aws cli as explained above.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [x] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

### CODE REVIEW

File: cmd/xl-v1-list-objects.go [hunk 1]
```diff
@@ -38,6 +38,10 @@ func listDirFactory(ctx context.Context, disks ...StorageAPI) ListDirFunc {
 				continue
 			}
 
+			if len(entries) == 0 {
```
Comment (vadmeste) [inline cmd/xl-v1-list-objects.go:41]: Shouldn't we take care of quorum thing there ? I mean what if the first disk does not have the object here but all other have ?

Comment (harshavardhana) [reply cmd/xl-v1-list-objects.go:41]: Quorum shall be validated at the top @vadmeste these entries get skipped invariably..

Comment (vadmeste) [reply cmd/xl-v1-list-objects.go:41]: > Quorum shall be validated at the top @vadmeste these entries get skipped invariably..

oh this is xl-v1 code, storage.Walk() is currently being used instead I guess

**Review by harshavardhana — APPROVED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#1e1c4ce8-80ba-4342-beb0-ac50c5a1895e) |
<a name="1e1c4ce8-80ba-4342-beb0-ac50c5a1895e"></a>
### 9129-37b94a2/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:30345
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp eaffed6bba46:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 9 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... FAILED in 6 minutes and 32 seconds
{
  "name": "aws-sdk-php",
  "function": "getBucketPolicy ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5e6c0f86f11a0"
  },
  "duration": "40582",
  "status": "FAIL",
  "error": "Error executing \"CreateBucket\" on \"http://minio-dev6.minio.io:30345/aws-sdk-php-5e6c0f86f11a0\"; AWS HTTP error: Client error: `PUT http://minio-dev6.minio.io:30345/aws-sdk-php-5e6c0f86f11a0` resulted in a `409 Conflict` response:\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>ContainerBeingDeleted</Code><Message>-&gt; github.com/Azure/azure-st (truncated...)\n ContainerBeingDeleted (client): -> github.com/Azure/azure-storage-blob-go/azblob.newStorageError, /home/mint-auto/mint-auto/9129-37b94a2/gopath/pkg/mod/github.com/!azure/azure-storage-blob-go@v0.8.0/azblob/zc_storage_error.go:42\n===== RESPONSE ERROR (ServiceCode=ContainerBeingDeleted) =====\nDescription=The specified container is being deleted. Try operation later.\nRequestId:e8017e25-e01e-00d3-468b-f9524b000000\nTime:2020-03-13T23:02:38.5877244Z, Details: \n   Code: ContainerBeingDeleted\n   PUT https://minioazure.blob.core.windows.net/aws-sdk-php-5e6c0f86f11a0?restype=container&timeout=16\n   Authorization: REDACTED\n   User-Agent: [APN/1.0 MinIO/1.0 MinIO/2020-03-13T22:42:57Z]\n   X-Ms-Client-Request-Id: [17a883de-806a-40b3-5e9d-e3ac163ebd25]\n   X-Ms-Date: [Fri, 13 Mar 2020 23:02:38 GMT]\n   X-Ms-Version: [2018-11-09]\n   --------------------------------------------------------------------------------\n   RESPONSE Status: 409 The specified container is being deleted. Try operation later.\n   Content-Length: [252]\n   Content-Type: [application/xml]\n   Date: [Fri, 13 Mar 2020 23:02:37 GMT]\n   Server: [Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0]\n   X-Ms-Error-Code: [ContainerBeingDeleted]\n   X-Ms-Request-Id: [e8017e25-e01e-00d3-468b-f9524b000000]\n   X-Ms-Version: [2018-11-09]\n\n\n - <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>ContainerBeingDeleted</Code><Message>-&gt; github.com/Azure/azure-storage-blob-go/azblob.newStorageError, /home/mint-auto/mint-auto/9129-37b94a2/gopath/pkg/mod/github.com/!azure/azure-storage-blob-go@v0.8.0/azblob/zc_storage_error.go:42&#xA;===== RESPONSE ERROR (ServiceCode=ContainerBeingDeleted) =====&#xA;Description=The specified container is being deleted. Try operation later.&#xA;RequestId:e8017e25-e01e-00d3-468b-f9524b000000&#xA;Time:2020-03-13T23:02:38.5877244Z, Details: &#xA;   Code: ContainerBeingDeleted&#xA;   PUT https://minioazure.blob.core.windows.net/aws-sdk-php-5e6c0f86f11a0?restype=container&amp;timeout=16&#xA;   Authorization: REDACTED&#xA;   User-Agent: [APN/1.0 MinIO/1.0 MinIO/2020-03-13T22:42:57Z]&#xA;   X-Ms-Client-Request-Id: [17a883de-806a-40b3-5e9d-e3ac163ebd25]&#xA;   X-Ms-Date: [Fri, 13 Mar 2020 23:02:38 GMT]&#xA;   X-Ms-Version: [2018-11-09]&#xA;   --------------------------------------------------------------------------------&#xA;   RESPONSE Status: 409 The specified container is being deleted. Try operation later.&#xA;   Content-Length: [252]&#xA;   Content-Type: [application/xml]&#xA;   Date: [Fri, 13 Mar 2020 23:02:37 GMT]&#xA;   Server: [Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0]&#xA;   X-Ms-Error-Code: [ContainerBeingDeleted]&#xA;   X-Ms-Request-Id: [e8017e25-e01e-00d3-468b-f9524b000000]&#xA;   X-Ms-Version: [2018-11-09]&#xA;&#xA;&#xA;</Message><BucketName>aws-sdk-php-5e6c0f86f11a0</BucketName><Resource>/aws-sdk-php-5e6c0f86f11a0</Resource><RequestId>15FBFF71E200B121</RequestId><HostId>73db9e73-eee3-417d-b8d1-641cde6b28ef</HostId></Error>"
}

Executed 2 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by vadmeste — APPROVED** [review]
LGTM

**Pull request merged**